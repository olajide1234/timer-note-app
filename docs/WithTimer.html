<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: WithTimer</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: WithTimer</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>WithTimer<span class="signature">()</span><span class="type-signature"></span></h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    

    
    <h4 class="name" id="WithTimer"><span class="type-signature"></span>new WithTimer<span class="signature">()</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    <p>Container to generate timer and pass to child components</p>
</div>













<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="views_WithTimer.jsx.html">views/WithTimer.jsx</a>, <a href="views_WithTimer.jsx.html#line91">line 91</a>
    </li></ul></dd>
    

    

    

    
</dl>



















    <h5>Example</h5>
    
    <pre class="prettyprint"><code>return (
  &lt;WithTimer />
)</code></pre>



    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="~resetFn"><span class="type-signature">(inner) </span>resetFn<span class="signature">(id)</span><span class="type-signature"> &rarr; {void}</span></h4>
    

    



<div class="description">
    <p>Calls function to reset timer</p>
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>id</code></td>
            

            <td class="type">
            
                
<span class="param-type">int</span>


            
            </td>

            

            

            <td class="description last"><p>ID of current set timeout</p></td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="views_WithTimer.jsx.html">views/WithTimer.jsx</a>, <a href="views_WithTimer.jsx.html#line120">line 120</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        


<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">void</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id="~startFn"><span class="type-signature">(inner) </span>startFn<span class="signature">()</span><span class="type-signature"> &rarr; {void}</span></h4>
    

    



<div class="description">
    <p>Calls function to start timer</p>
</div>













<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="views_WithTimer.jsx.html">views/WithTimer.jsx</a>, <a href="views_WithTimer.jsx.html#line104">line 104</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        


<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">void</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id="~stopFn"><span class="type-signature">(inner) </span>stopFn<span class="signature">(id)</span><span class="type-signature"> &rarr; {void}</span></h4>
    

    



<div class="description">
    <p>Function to stop timer</p>
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>id</code></td>
            

            <td class="type">
            
                
<span class="param-type">int</span>


            
            </td>

            

            

            <td class="description last"><p>ID of current set timeout</p></td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="views_WithTimer.jsx.html">views/WithTimer.jsx</a>, <a href="views_WithTimer.jsx.html#line112">line 112</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        


<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">void</span>


    </dd>
</dl>

    





        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="AddNote.html">AddNote</a></li><li><a href="Timer.html">Timer</a></li><li><a href="ViewNotes.html">ViewNotes</a></li><li><a href="ViewSingleNote.html">ViewSingleNote</a></li><li><a href="WithTimer.html">WithTimer</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc/jsdoc">JSDoc 3.6.3</a> on Tue Jan 21 2020 00:55:03 GMT+0100 (West Africa Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>