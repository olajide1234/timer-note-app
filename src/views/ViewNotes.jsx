import React from 'react'
import { ViewSingleNote } from '../components'
import { useSelector, useDispatch } from "react-redux";

/**
 * Container to fetch and display all notes
 *
 * @component
 * @example
 * return (
 *   <ViewNotes />
 * )
 */
export function ViewNotes() {
  const dispatch = useDispatch();
  const notes = useSelector(state => state);

  /**
   * Function to delete a note 
   *
   * @param   {int} id Id of note to be deleted
   * @returns  {void}
   */
  const deleteFn = (id) => dispatch({ type: 'DELETE_NOTE', deleteId: id })

  return (
    <div>
      <div className='headerText'>View and Delete Notes</div>
      <div className='noteContainer'>
        {notes.length < 1 ? 'Add notes to see them appear here' : [...notes].reverse().map((note, i) => <ViewSingleNote key={i} id={note.id} time={note.time} text={note.text} deleteFn={deleteFn} />)}
      </div>
    </div>
  );
}
