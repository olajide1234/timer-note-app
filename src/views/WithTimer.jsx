import React, { Fragment, useState } from 'react'
import { AddNote, Timer } from '../components'
import { Col } from 'antd';

var hours = 0;
var mins = 0;
var seconds = 0;

/**
 * Function to start timer
 * Because JS does not pass-by-reference and we need to persist changes to hrs, mins and secs, we don't pass these as arguments.
 * We also simulate overloading on this function for testing
 * @param   {function} updateTimerHrs Function to update hour of timer
 * @param   {function} updateTimerMins Function to update minutes of timer
 * @param   {function} updateTimerSecs Function to update seconds of timer
 * @param   {function} updateTimerStopId Function to id of timer for stopping
 * @param   {int} h Current hour for overloading/testing
 * @param   {int} m Current minute for overloading/testing
 * @param   {int} s Current secoond for overloading/testing
 * @returns  {void} 
 */

export const startTimer = (updateTimerHrs, updateTimerMins, updateTimerSecs, updateTimerStopId, h, m, s) => {
  var stopId = setTimeout(() => {
    s ? s++ : seconds++;

    if (seconds > 59 || s > 59) {
      s ? s = 0 : seconds = 0; m ? m++ : mins++;
      if (mins > 59 || m > 59) {
        m ? m = 0 : mins = 0; h ? h++ : hours++;
        if (h) {
          if (h < 10) { updateTimerHrs(`0${hours}:`); } else { updateTimerHrs(`${hours}:`); }
        } else {
          if (hours < 10) { updateTimerHrs(`0${hours}:`); } else { updateTimerHrs(`${hours}:`); }
        }
      }

      if (m) {
        if (m < 10) {
          updateTimerMins(`0${mins}:`);
        }
        else {
          updateTimerMins(`${mins}:`);
        }
      } else {
        if (mins < 10) {
          updateTimerMins(`0${mins}:`);
        }
        else {
          updateTimerMins(`${mins}:`);
        }
      }

    }
    if (seconds < 10 || s < 10) {
      updateTimerSecs(`0${seconds}`);
    } else {
      updateTimerSecs(`${seconds}`);
    }
    updateTimerStopId(stopId)
    startTimer(updateTimerHrs, updateTimerMins, updateTimerSecs, updateTimerStopId);
  }, 1000);
}

/**
 * Resets timer
 *
 * @param   {int} id ID of current set timeout
 * @param   {function} updateTimerHrs Function to update hour of timer
 * @param   {function} updateTimerMins Function to update minutes of timer
 * @param   {function} updateTimerSecs Function to update seconds of timer
 * @returns  {void}
 */
export const resetTimer = (id, updateTimerHrs, updateTimerMins, updateTimerSecs) => {
  clearTimeout(id);
  hours = 0; mins = 0; seconds = 0;
  updateTimerSecs('00');
  updateTimerMins('00:');
  updateTimerHrs('00:');
}

/**
 * Container to generate timer and pass to child components
 *
 * @component
 * @example
 * return (
 *   <WithTimer />
 * )
 */
export function WithTimer() {

  const [timerStopId, updateTimerStopId] = useState(0);
  const [timerSecs, updateTimerSecs] = useState('00');
  const [timerMins, updateTimerMins] = useState('00:');
  const [timerHrs, updateTimerHrs] = useState('00:');
  const [timerPause, setTimerPause] = useState(false);

  /**
   * Calls function to start timer
   *
   * @returns  {void} 
   */
  const startFn = () => { startTimer(updateTimerHrs, updateTimerMins, updateTimerSecs, updateTimerStopId); setTimerPause(false) }

  /**
   * Function to stop timer
   *
   * @param   {int} id ID of current set timeout
   * @returns  {void}
   */
  const stopFn = (id) => { clearTimeout(id); setTimerPause(true) };

  /**
   * Calls function to reset timer
   *
   * @param   {int} id ID of current set timeout
   * @returns  {void}
   */
  const resetFn = (id) => { setTimerPause(false); resetTimer(id, updateTimerHrs, updateTimerMins, updateTimerSecs); }

  return (
    <Fragment>
      <Col>
        <Timer
          time={`${timerHrs}${timerMins}${timerSecs}`}
          startFn={startFn}
          stopFn={stopFn}
          resetFn={resetFn}
          currentTimerStopId={timerStopId + 1}
          timerPause={timerPause}
        />
      </Col>
      <Col><AddNote time={`${timerHrs}${timerMins}${timerSecs}`} stopFn={stopFn} currentTimerStopId={timerStopId + 1} /></Col>
    </Fragment>
  );
}
