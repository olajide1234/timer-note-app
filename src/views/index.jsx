export { ViewNotes } from './ViewNotes';
export { WithTimer } from './WithTimer'
