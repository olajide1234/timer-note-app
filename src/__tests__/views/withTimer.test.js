import React from 'react';
import { Provider } from 'react-redux'
import { WithTimer } from '../../views';
import { startTimer, resetTimer } from '../../views/WithTimer';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store'

const mockStore = configureMockStore();

it('should render timer correctly', () => {
  const store = mockStore([{ id: 1, time: '12:12:12', text: 'Here is a sample note. Add more notes to see your note above this.' }]);

  const tree = renderer.create(
    <Provider store={store}>
      <WithTimer />
    </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should update timer correctly', () => {
  const mockFn1 = jest.fn();
  const mockFn2 = jest.fn();
  const mockFn3 = jest.fn();
  const mockFn4 = jest.fn();

  jest.useFakeTimers();

  startTimer(mockFn1, mockFn2, mockFn3, mockFn4, 1, 11, 60);
  expect(setTimeout).toHaveBeenCalledTimes(1);

  jest.advanceTimersByTime(1000);
  startTimer(mockFn1, mockFn2, mockFn3, mockFn4, 1, 60, 60);
  expect(mockFn2).toHaveBeenCalled();
  jest.advanceTimersByTime(1000);

  jest.advanceTimersByTime(1000);
  startTimer(mockFn1, mockFn2, mockFn3, mockFn4, 12, 60, 60)
  expect(mockFn1).toHaveBeenCalled();
  jest.advanceTimersByTime(1000);

  jest.advanceTimersByTime(1000);
  startTimer(mockFn1, mockFn2, mockFn3, mockFn4, 0, 6, 59)
  expect(mockFn2).toHaveBeenCalled();
  jest.advanceTimersByTime(1000);
});

it('should reset timer', () => {
  const mockFn = jest.fn();
  expect(resetTimer(2, mockFn, mockFn, mockFn)).toMatchSnapshot();
});
