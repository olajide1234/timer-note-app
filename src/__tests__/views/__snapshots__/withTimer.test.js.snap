// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render timer correctly 1`] = `
Array [
  <div
    className="ant-col"
  >
    <div>
      <div
        className="headerText"
      >
        Control Timer
      </div>
      <div
        className="timerFace"
      >
        <div>
          00:00:00
        </div>
      </div>
      <div
        className="ant-row-flex ant-row-flex-space-between"
        style={Object {}}
      >
        <div
          className="ant-col"
          style={Object {}}
        >
          <button
            className="ant-btn"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <span>
              Start
            </span>
          </button>
        </div>
        <div
          className="ant-col"
          style={Object {}}
        >
          <button
            className="ant-btn"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <span>
              Pause
            </span>
          </button>
        </div>
        <div
          className="ant-col"
          style={Object {}}
        >
          <button
            className="ant-btn"
            onClick={[Function]}
            type="button"
          >
            <span>
              Reset
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="ant-col"
  >
    <div>
      <div
        className="headerText"
      >
        Add Notes
      </div>
      <div
        className="smallTopMargin"
      >
        <form
          className="ant-form ant-form-horizontal"
          onSubmit={[Function]}
        >
          <div
            className="ant-row ant-form-item"
            style={Object {}}
          >
            <div
              className="ant-col ant-form-item-control-wrapper"
              style={Object {}}
            >
              <div
                className="ant-form-item-control"
              >
                <span
                  className="ant-form-item-children"
                >
                  <span
                    className="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
                  >
                    <textarea
                      className="ant-input"
                      cols="33"
                      onChange={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Jot your note here"
                      rows="5"
                      style={Object {}}
                      value=""
                    />
                  </span>
                </span>
              </div>
            </div>
          </div>
          <button
            className="ant-btn ant-btn-primary"
            onClick={[Function]}
            type="submit"
          >
            <span>
              Save note
            </span>
          </button>
        </form>
      </div>
    </div>
  </div>,
]
`;

exports[`should reset timer 1`] = `undefined`;
